File: src/types/caldav.ts
/**
 * CalDAV accounts — the wire contract with `src-tauri/src/schema/caldav.rs`.
 *
 * Same shapes, same names, same optionality. `docs/caldav_plan.md` is the
 * design; the short version of what these types are *for*:
 *
 * An account is the **mail-account shape**, not the `source_url` shape — a
 * user-typed server, a password that is never persisted without an explicit
 * opt-in, a background interval, several accounts. `Calendar.source_url` (an
 * anonymous ICS feed) stays exactly as it was and is still the right tool for
 * a read-only link someone shared with you.
 *
 * The **events and tasks a sync produces are ordinary `CalendarEvent`s and
 * `CalendarTask`s** in `calendar.json`, carrying `caldav_href`/`caldav_etag`.
 * Nothing that renders a calendar knows this feature exists.
 */

import type { CalendarEvent, CalendarTask } from "./index";

/** One subscribed collection on an account. */
export interface CalDavCalendarRef {
  /** The collection's absolute URL — **the** stable key. */
  href: string;
  /** The `Calendar.id` in `calendar.json` this collection syncs into. */
  calendar_id: string;
  display_name: string;
  /** Last-seen `getctag`: the cheap "did anything change at all" check. */
  ctag: string;
  /** Last-seen RFC 6578 token, when the server offers one. */
  sync_token?: string | null;
  /** `VEVENT` / `VTODO`. Empty means the server did not say. */
  components?: string[];
  /** **The server's own answer**, from its `current-user-privilege-set`: a
   *  collection someone else shared with you is legitimately read-only to you.
   *  Half of the push gate — `CalDavAccount.allow_write` is the other half, and
   *  the server's 403 is the only one that is actually enforcement. */
  read_only: boolean;
  /** Local stamp of the last successful sync (`"YYYY-MM-DDTHH:MM"`). */
  last_sync?: string;
}

export interface CalDavAccount {
  id: string;
  label: string;
  /** What discovery resolved, or what the user pasted. Never inferred from
   *  anything — least of all from an email address. */
  base_url: string;
  user: string;
  /** Opt-in, **default false**. */
  save_password: boolean;
  /** Minutes between background syncs; `0`/absent means manual-only. */
  sync_interval_min?: number | null;
  /** **Two-way sync — opt-in, default false.** The plan left "is write access
   *  even wanted against an institutional calendar?" open; this is that question
   *  asked rather than answered on the user's behalf. A push bug against a
   *  shared work calendar has a materially bigger blast radius than the same bug
   *  against a self-hosted one, and only the account's owner knows which they
   *  are looking at. */
  allow_write?: boolean;
  calendars: CalDavCalendarRef[];
}

/** What one write did — see `schema/caldav.rs`'s `CalDavWrite`. */
export interface CalDavWrite {
  href: string;
  /** The validator for the next conditional write. Empty means the server named
   *  none, and the next write waits for a sync rather than going unconditional. */
  etag: string;
  /** The server refused: the resource changed elsewhere since `etag` was read
   *  (412), or something already exists where a create aimed. A **result**, not
   *  an error — the user has a decision to make and needs it as a value. */
  conflict: boolean;
  /** A delete found nothing there. The intended end state is the actual one. */
  gone: boolean;
}

/** What the keychain actually did — never collapsed to a bare account. */
export interface CalDavAccountSaved {
  account: CalDavAccount;
  saved: boolean;
  save_error?: string | null;
}

export interface CalDavPasswordState {
  has_saved: boolean;
  /** `"unlocked" | "locked" | "unavailable"` — the shared `KeyringState`. */
  keyring: "unlocked" | "locked" | "unavailable";
}

/** One calendar collection discovery found. */
export interface CalDavCollection {
  href: string;
  display_name: string;
  /** `calendar-color` as the server spells it (`#rrggbb` or `#rrggbbaa`). */
  color: string;
  ctag: string;
  sync_token?: string | null;
  components: string[];
  /** The server reported privileges and none was a write privilege. */
  read_only: boolean;
}

/** One resource as the server handed it over — iCalendar text, unparsed. */
export interface CalDavResource {
  href: string;
  etag: string;
  data: string;
}

/** What one fetch found. */
export interface CalDavChanges {
  resources: CalDavResource[];
  /** Hrefs the server reported **gone**. The only signal that deletes a task. */
  removed: string[];
  sync_token?: string | null;
  ctag: string;
  /** A `sync-collection` reply: an absent href means *unchanged*, not deleted. */
  incremental: boolean;
  /** The ctag matched: nothing was fetched, nothing needs applying. */
  unchanged: boolean;
}

/** One resource after `lib/ics.ts` has parsed it — what `caldav_apply` takes. */
export interface CalDavParsed {
  href: string;
  etag: string;
  events: CalendarEvent[];
  tasks: CalendarTask[];
}

File: src-tauri/src/schema/caldav.rs
//! Serde structs for CalDAV accounts (`docs/caldav_plan.md`).
//!
//! Two jobs, the same two `schema::mail` has:
//!
//! 1. **The wire contract** with `src/types/caldav.ts` — snake_case, same field
//!    names, same optionality.
//! 2. **`accounts.json`.** [`CalDavAccounts`] is the on-disk store under
//!    `~/.local/share/eldrun/caldav/accounts.json`. It carries **no secret** —
//!    passwords live in the OS keychain via `services::remote_credentials`,
//!    keyed by server target (`commands::caldav::account_key`).
//!
//! **Why this is a second file and not a few keys on `Calendar`.** `source_url`
//! (the anonymous ICS-feed case) rides `Calendar`'s `extra` flatten because it
//! names nothing that needs protecting and carries no cursor. A CalDAV account
//! carries a username, a per-collection sync cursor, and a credential in the
//! keychain — that is account state, not calendar-display state, and putting it
//! in `calendar.json` would mean the file every calendar tab reads on mount (and
//! that gets exported and inspected alongside a calendar) also carries sync
//! bookkeeping nothing there needs. The mail client draws the same line between
//! `MailAccounts` and the message index.
//!
//! The **events and tasks themselves stay in `calendar.json`** as ordinary
//! `CalendarEvent`/`CalendarTask` rows, each carrying `caldav_href` and
//! `caldav_etag` in its own `extra` flatten. That is what lets every existing
//! consumer — the grid, month view, agenda, alarms, ICS export, the to-do board
//! — render CalDAV-sourced rows with zero changes: they already do not know or
//! care where a row came from.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::services::remote_credentials::KeyringState;

// ── Accounts ────────────────────────────────────────────────────────────────

/// One subscribed collection on an account.
///
/// `href` is the collection's own URL and is **the** stable key: the account's
/// label can be renamed, the local `Calendar` can be recolored, but the href is
/// what the server recognizes.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct CalDavCalendarRef {
    /// The collection's absolute URL.
    pub href: String,
    /// The `Calendar.id` in `calendar.json` this collection syncs into.
    pub calendar_id: String,
    #[serde(default)]
    pub display_name: String,
    /// Last-seen `getctag` — the cheap "did anything change at all" check that
    /// lets a scheduled sync skip the expensive report entirely.
    #[serde(default)]
    pub ctag: String,
    /// Last-seen RFC 6578 `sync-token`, when the server offers one. `None`
    /// means the ctag-gated full refetch path, which every server supports.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sync_token: Option<String>,
    /// Which components the collection holds (`VEVENT`, `VTODO`). Empty means
    /// the server did not say, and both are asked for.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub components: Vec<String>,
    /// **True in Phases 1–2, always.** The local calendar is server-authoritative
    /// and has no push path yet; the field exists so the day push lands, the
    /// answer is per-collection and comes from the server's own
    /// `current-user-privilege-set` rather than from "the feature shipped".
    #[serde(default = "default_true")]
    pub read_only: bool,
    /// UTC-ish local stamp of the last successful sync, as the frontend minted
    /// it (this crate pulls in no time crate — same reasoning as
    /// `CalendarTask::created`). Empty until the first success.
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub last_sync: String,
    #[serde(flatten, default)]
    pub extra: HashMap<String, Value>,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct CalDavAccount {
    pub id: String,
    #[serde(default)]
    pub label: String,
    /// What discovery resolved, or what the user pasted directly. Never
    /// defaulted from anything: a university-affiliated user can hold two
    /// unrelated mailboxes on two unrelated systems, so which account to sync is
    /// theirs to type — exactly like a mail account's IMAP host.
    pub base_url: String,
    pub user: String,
    /// Opt-in, **default false** — the standing no-password-storage rule.
    /// False means the password lives in the backend's in-memory map for this
    /// session only.
    #[serde(default)]
    pub save_password: bool,
    /// Minutes between background syncs. `None` or `0` means manual-only.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sync_interval_min: Option<u32>,
    #[serde(default)]
    pub calendars: Vec<CalDavCalendarRef>,
    #[serde(flatten, default)]
    pub extra: HashMap<String, Value>,
}

/// `~/.local/share/eldrun/caldav/accounts.json`.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CalDavAccounts {
    #[serde(default)]
    pub version: u32,
    #[serde(default)]
    pub accounts: Vec<CalDavAccount>,
    #[serde(flatten, default)]
    pub extra: HashMap<String, Value>,
}

/// Current `accounts.json` version. Bumped only for a shape change a reader has
/// to branch on; additive fields ride the `extra` catch-all.
pub const ACCOUNTS_VERSION: u32 = 1;

// ── Command results ─────────────────────────────────────────────────────────

/// What the keychain **actually did**, never collapsed to a bare account: a
/// write that silently failed is how a user loses a password they think is
/// saved. Identical in shape to `MailAccountSaved` because it answers the same
/// generic question.
#[derive(Debug, Clone, Serialize)]
pub struct CalDavAccountSaved {
    pub account: CalDavAccount,
    pub saved: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub save_error: Option<String>,
}

/// Whether a password is saved for an account — and whether the store could be
/// read at all, because a locked collection answers every lookup with "nothing
/// saved" and reading that as absence is what un-saves credentials.
#[derive(Debug, Clone, Serialize)]
pub struct CalDavPasswordState {
    pub has_saved: bool,
    pub keyring: KeyringState,
}

// ── Protocol shapes ─────────────────────────────────────────────────────────

/// One calendar collection discovery found. The account dialog lists these as
/// checkboxes; subscribing turns the ticked ones into [`CalDavCalendarRef`]s.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct CalDavCollection {
    pub href: String,
    #[serde(default)]
    pub display_name: String,
    /// `calendar-color`, as the server spells it (`#rrggbb` or `#rrggbbaa`).
    #[serde(default)]
    pub color: String,
    #[serde(default)]
    pub ctag: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sync_token: Option<String>,
    #[serde(default)]
    pub components: Vec<String>,
    /// The server reported privileges and none of them was a write privilege.
    /// A server that reports none at all leaves this `false` — unknown, not a
    /// claim of writability.
    #[serde(default)]
    pub read_only: bool,
}

/// One resource (one VEVENT or VTODO, by universal convention) as the server
/// handed it over: its URL, its ETag, and its iCalendar text **unparsed**.
///
/// The text goes to the frontend and through `src/lib/ics.ts`, which is the one
/// parser here that understands folding, escaping, `RRULE` and `VALARM` — the
/// same handoff `calendar_fetch_ics` already makes.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct CalDavResource {
    pub href: String,
    #[serde(default)]
    pub etag: String,
    #[serde(default)]
    pub data: String,
}

/// One resource **after** the frontend has parsed its iCalendar text.
///
/// The round trip is deliberate and mirrors the ICS path exactly: the backend
/// fetches ([`CalDavResource`]), `src/lib/ics.ts` parses, and the parsed rows
/// come back here to be reconciled and written. The alternative — a second
/// iCalendar parser in Rust — would mean two implementations of folding,
/// escaping, `RRULE` and `VALARM` that could disagree about the same feed.
///
/// The resource identity travels **with** the group rather than being poked
/// into each row's `extra` by the frontend, so the one place that decides what
/// `caldav_href` and `caldav_etag` mean is `commands::calendar`.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct CalDavParsed {
    pub href: String,
    #[serde(default)]
    pub etag: String,
    #[serde(default)]
    pub events: Vec<crate::schema::calendar::CalendarEvent>,
    #[serde(default)]
    pub tasks: Vec<crate::schema::calendar::CalendarTask>,
}

/// What one fetch of a collection found.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct CalDavChanges {
    pub resources: Vec<CalDavResource>,
    /// Hrefs the server reported as **gone** (RFC 6578's `404` stubs, or a
    /// `404` from a targeted multiget). This is the only signal that deletes a
    /// task: absence from a listing is not, because some servers stop returning
    /// completed VTODOs by default filter and "deleted" and "filtered out" would
    /// be indistinguishable.
    pub removed: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sync_token: Option<String>,
    #[serde(default)]
    pub ctag: String,
    /// This was an incremental (`sync-collection`) report, so a local row whose
    /// href is *absent* means "unchanged", never "deleted".
    #[serde(default)]
    pub incremental: bool,
    /// The ctag matched the stored one: nothing was fetched and nothing needs
    /// applying. The whole point of the cheap check.
    #[serde(default)]
    pub unchanged: bool,
}
